torch/csrc/autograd/compiled_autograd.h

```diff
@@ -0,0 +1,583 @@
+  }
+  void collect(const c10::SymInt& t) {
+    _compiler.add_size_input(t);
+  }
+  template <typename T>
+  void collect(const std::vector<T>& t) {
+    collect_size(t.size());
+    for (size_t i = 0; i < t.size(); ++i) {
+      collect(t[i]);
+    }
+  }
+  template <typename T>
+  void collect(const c10::ArrayRef<T>& t) {
+    collect_size(t.size());
+    for (size_t i = 0; i < t.size(); ++i) {
+      collect(t[i]);
+    }
+  }
+  template <typename T>
+  void collect(const c10::OptionalArray<T>& t) {
+    collect(t.list);
+  }
+  template <typename T>
+  void collect(const c10::optional<T>& t) {
+    specialize_on_bytes(t.has_value());
+    if (t.has_value()) {
+      collect(*t);
+    }
+  }
+  template <typename A, typename B>
+  void collect(const std::pair<A, B>& t) {
+    collect(t.first);
+    collect(t.second);
+  }
+  void collect(const c10::ScalarType& t) {
+    specialize_on_bytes(t);
+  }
+  void collect(const c10::Scalar& t) {
+    auto type = t.type();
+    specialize_on_bytes(type);
+    if (type == c10::ScalarType::Double) {
+      collect(t.toDouble());
+    } else if (type == c10::ScalarType::Long) {
+      collect(t.toLong());
+    } else if (type == c10::ScalarType::Bool) {
+      collect(t.toBool());
+    } else if (type == c10::ScalarType::ComplexDouble) {
+      auto c = t.toComplexDouble();
+      collect(c.real());
+      collect(c.imag());
+    } else {
+      TORCH_CHECK(false);
+    }
+  }
+  void collect(const c10::TensorOptions& t) {
+    // TODO(jansel): are there any pointers in this type we shouldn't memcmp
+    // I think this one is wrong.... should fix it
+    specialize_on_bytes(t);
+  }
+  void collect(const at::TensorGeometry& t) {
+    collect(t.sym_sizes());
+    collect(t.sym_strides());
+    collect(t.sym_storage_offset());
+  }
+  void collect(torch::autograd::generated::TypeAndSize& t);
+  void collect(const OutputRef& t) {
+    // +1 is for undefined encoded as -1
+    collect_size(t.node_id + 1);
+    collect_size(t.index + 1);
+  }
+  void collect(const NodeCall& t) {
+    collect(t.input_refs);
+    collect(t.graph_output);
+  }
+  void collect(int8_t t) {
+    specialize_on_bytes(t);
+  }
+  void collect(int16_t t) {
+    specialize_on_bytes(t);
+  }
+  void collect(int32_t t) {
+    specialize_on_bytes(t);
+  }
+  void collect(int64_t t) {
+    specialize_on_bytes(t);
+  }
+  void collect(uint8_t t) {
+    specialize_on_bytes(t);
+  }
+  void collect(uint16_t t) {
+    specialize_on_bytes(t);
+  }
+  void collect(uint32_t t) {
+    specialize_on_bytes(t);
+  }
+  void collect(uint64_t t) {
+    specialize_on_bytes(t);
+  }
+  void collect(bool t) {
+    specialize_on_bytes(t);
+  }
+  void collect(float t) {
+    specialize_on_bytes(t);
+  }
+  void collect(double t) {
+    specialize_on_bytes(t);
+  }
+
+  void set_grad_target(const at::Tensor& tensor) {
+    collect(_compiler.accumulate_grad);
+    if (_compiler.accumulate_grad)
+      _compiler.add_set_grad_target(tensor);
+  }
```

Comment:
Not entirely sure what this is doing

Reply:
It is recording the Tensor object that we later do:
```
x.mutable_grad() = ...
```
on (based on the return value of the FX graph).